### PR TITLE
Solution for #20: Windows process argument quoting

### PR DESCRIPTION
**Pull request by patrickbkr**

### CODE REVIEW

**Comment by AlexDaniel** [comment]
@jnthn please feel free to reassign this PR.

File: solutions/language/Windows-process-argument-quoting.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+- Work in the common case without the user having to know about any of this
+- Give enough control to the user to call any program the way he wants
+- Make sure the changes don't make the API more complex for the use cases not affected by this
+
+
+Solution
+--------
+
+`sub run()`, `Proc.Spawn()` and `Proc::Async.new()` are extended with a new argument `$:win-verbatim-args` defaulting to `False`. If the argument is left off or set to `False`, the arguments are quoted according to the Microsoft convention. This is identical to Rakus previous behavior and thus backwards compatible.
```
Comment (jnthn) [inline solutions/language/Windows-process-argument-quoting.md:20]: Maybe `:$win-verbatim-args`, since that is how it shows in the signature?

**Comment by AlexDaniel** [comment]
Also, this PR seems to come with drafts for:
* MoarVM https://github.com/MoarVM/MoarVM/pull/1265
* NQP https://github.com/Raku/nqp/pull/610
* Rakudo https://github.com/rakudo/rakudo/pull/3576
* Docs https://github.com/Raku/doc/pull/3347
* Spec https://github.com/Raku/roast/pull/636

Please edit this comment if there's anything else. It's good to know how the implementation is going to change.

File: solutions/language/Windows-process-argument-quoting.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+Other serialization strategies can be implemented in ecosystem modules.
+
+The name `win-verbatim-args` is named similar to a matching flag in `libuv` and `node.js`. This increases the recognition factor.
+
+
+Implementation in Rakudo
+------------------------
+
+The serialization of the arguments happens in Rakudo. Backends are expected to not touch the single arument string they receive from Rakudo on Windows in any way.
```
Comment (jnthn) [inline solutions/language/Windows-process-argument-quoting.md:32]: This is probably good for backend consistency. I'm an iota concerned about reproducing a quoting algorithm, which has security implications, but relieved that there's tests covering this.

File: solutions/language/Windows-process-argument-quoting.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+- Documentation of the Windows API function to parse command line arguments: [CommandLineToArgv()](https://docs.microsoft.com/en-us/windows/desktop/api/shellapi/nf-shellapi-commandlinetoargvw)
+
+
+### What other languages do
+
+[libuv](https://github.com/libuv/libuv/blob/c5593b51dc98715f7f32a919301b5801ebf1a8ce/src/win/process.c#L593) quotes arguments compatible to `CommandLineToArgv()`. When passing the `UV_PROCESS_WINDOWS_VERBATIM_ARGUMENTS` argument this quoting is disabled and the arguments are just joined with a space.
+
+[Node.js](https://github.com/nodejs/node/blob/382e859afc7e66600dccfadd4125088444e063c3/lib/child_process.js#L486) exposes a `shell` flag which wraps the call in `cmd.exe /d /s /c "<command>"` by hand and enables `UV_PROCESS_WINDOWS_VERBATIM_ARGUMENTS`.
+
+[Java](https://codewhitesec.blogspot.com/2016/02/java-and-command-line-injections-in-windows.html)s `ProcessBuilder` has a very sophisticated and complex processing of arguments. e.g. it detects whether the user called a `.bat` file or `.cmd` file and acts on that.
```
Comment (jnthn) [inline solutions/language/Windows-process-argument-quoting.md:52]: I'm glad we didn't go this way.

File: solutions/language/Windows-process-argument-quoting.md (hunk 1 above)
Comment (patrickbkr) [reply solutions/language/Windows-process-argument-quoting.md]: Corrected

**Comment by AlexDaniel** [comment]
@patrickbkr so is this PR ready now?

**Comment by patrickbkr** [comment]
@AlexDaniel: I think this PR is ready. But IIUC it still needs to go through the review period.

**Comment by patrickbkr** [comment]
@AlexDaniel: 13 days have passed. I think the two week review period is meant for people to look at this and give feedback. I think barely anyone looked at this. What shall we do?

**Comment by lizmat** [comment]
I think (I hope I'm wrong) that @AlexDaniel is taking a break from Raku.

As an admin I will merge this tomorrow: I don't see a point in adding reviewer at this point in time, as that would only delay the process.

**Comment by patrickbkr** [comment]
I'm fine with a review period. But someone needs to add the reviewers...

**Comment by AlexDaniel** [comment]
> I'm fine with a review period. But someone needs to add the reviewers...

I was adding reviewers simply to generate email spam so that the reviewers are notified. You can also do that by simply tagging their names (by copying the list from https://github.com/Raku/problem-solving#reviewers). I also hope that reviewers will check the PRs at least once a month, or ideally once every 14 weeks.

> @AlexDaniel is taking a break from Raku.

That is correct.

**Pull request merged**